File: examples/testreactor_pso/main_pso.cpp
#include<iostream>
#include "../../core/parameter_estimation.h"
#include "../../core/parameters.h"
#include <string>
#include <filesystem>

int main(int argc, char *argv[]){

	std::vector<std::string> names={"K_s_aceto_Methanogenesis","k_m_aceto_Methanogenesis","k_dec_MO_acetoM","k_dec_MO_hydroM","k_dec_MO_ButyricValeric","k_dec_MO_Propionic","k_dec_MO_AcidoMS","k_dec_MO_AcidoAA","k_dec_MO_AcidoLCFA"};
		
	//std::string dir="/home/devanshr/Programs/ug4/plugins/ConstrainedOptimization/examples/testreactor_pso/";
	std::string dir=std::filesystem::current_path().string();
	

	std::vector<co::EFloat64> bounds= {co::EFloat64(1e-6),co::EFloat64(0.5),co::EFloat64(0.1),co::EFloat64(0.5)};
	for(int i = 0; i<14;i+=2){
		bounds.push_back(co::EFloat64(1e-06));
		bounds.push_back(co::EFloat64(1e-01));
	}


	co::PSOOptions options;
	options.set_max_iterations(20);
	options.set_n_particles(8);
	options.set_n_groups(2);
	co::BiogasEvaluation<co::EFloat64,co::ConfigComputation::Local, co::ConfigOutput::File> evaluator(dir,"subset_target.lua", "subset_sim.lua");
	co::ParticleSwarmOptimizer<co::BiogasEvaluation<co::EFloat64,co::ConfigComputation::Local, co::ConfigOutput::File>> pso(options,evaluator);
	co::EVarManager<co::EFloat64> estimated_parameters;
	pso.run(estimated_parameters,names,bounds);
}

File: examples/testreactor_pso/mainpso.lua
--Only parameters added to this container will be considered by the
--estimation procedure
manager=VarDescriptor64()

--Defining parameters
--The first number is the initial parameter value and the following
--two numbers give lower and upper bounds that must not be exceeded by
--the optimization procedure

manager:add("k_m_aceto_Methanogenesis",EFloat64(1e-6),EFloat64(0.5))
manager:add("K_s_aceto_Methanogenesis",EFloat64(0.1),EFloat64(0.5))
manager:add("k_dec",EFloat64(8e-04),EFloat64(0.1))

--When the estimation is finished, new parameter values can be accessed
--from Lua via this container
estimated_parameters=EVar64Manager()

--Running the Particle Swarm Optimization Algorithm
n_particles=12
n_groups=2
max_iterations=20 --maximum iterations of the PSO algorithm if no convergence is reached beforehand
RunPSO_BiogasEval("/home/devanshr/Programs/ug4/plugins/ConstrainedOptimization/examples/testreactor_pso",manager,estimated_parameters,n_particles,n_groups,max_iterations)

--[[Generate a lua table with the parameters.
Note: The paramters are converted to double which means they lose
their accumulated error bounds.
--]]
tab={}
for k=0,estimated_parameters:len()-1 do
			
			name=estimated_parameters:get_name(k)
			value=estimated_parameters:get_param(k):get_value_as_double()
			tab[name]=value
end
print("Parameter values converted into lua table:")
print(tab)
print("Execution completed") 
